File: 프로그래머스/3/43162. 네트워크/네트워크.js
function solution (n, computers) {
    const graph = new Graph();
    let networkCnt = 0;
    
    for (let i = 0; i < computers.length; i++) {
        for (let j = 0; j < computers.length; j++) {
            if (computers[i][j] === 1) {
                graph.addVertex(i + 1);
                graph.addVertex(j + 1);
                if (i !== j) {
                    graph.addEdge(i + 1, j + 1);
                }
            }
        }
    }
    
    function bfs (start, visited) {
        let cnt = 0;
        const queue = [start];
        visited.add(start);
        
        while (queue.length > 0) {
            const elem = queue.shift();
            graph.adjacencyList[elem].forEach((vertex) => {
                if (!visited.has(vertex)) {
                    visited.add(vertex)
                    queue.push(vertex)
                }
            })
        }
    }
    
    const visited = new Set();
    
    for (let i = 1; i <= n; i++) { //모든 정점 기준으로 순회하되 방문한 적 있으면 스킵
        if (visited.size === n) break;
        if (!visited.has(i)) {
            bfs(i, visited);
            networkCnt++;
        }
    }
    
    return networkCnt;
}

class Graph {
    constructor() {
        this.adjacencyList = {};
    }
    addVertex(vertex) {
        if (!this.adjacencyList[vertex]) this.adjacencyList[vertex] = new Set();
    }
    addEdge(v1, v2) {
        this.adjacencyList[v1].add(v2);
        this.adjacencyList[v2].add(v1);
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def bfs(graph, start, visited):
    queue = deque([start])
    visited[start] = True
    
    while queue:
        elem = queue.popleft()
        
        for adj in graph[elem]:
            if not visited[adj]:
                visited[adj] = True
                queue.append(adj)

        


def solution(n, computers):
    graph = {}
    answer = 0
    
    for idx in range(1, n + 1):
        graph[idx] = []
        
    # 그래프 초기화
    for i in range(n):
        for j in range(n):
            if i != j and computers[i][j] == 1:
                if j + 1 not in graph[i + 1]:
                    graph[i + 1].append(j + 1)
                if i + 1 not in graph[j + 1]:
                    graph[j + 1].append(i + 1)
    
    visited = [False] * (n + 1)
        
    for idx in range(1, n + 1):
        if not visited[idx]:
            visited[idx] = True
            bfs(graph, idx, visited)
            answer += 1
    return answer